OsmAnd/src/net/osmand/plus/views/AsynchronousResampler.java

```diff
@@ -0,0 +1,282 @@
+
+    private List<GPXUtilities.WptPt> culled = null;
+
+
+    public AsynchronousResampler(Renderable.RenderType rt, OsmandMapTileView view,
+                                 Renderable.RenderableSegment rs, double param1, double param2) {
+        this.view = view;
+        this.rs = rs;
+        this.renderType = rt;
+
+        this.param1 = param1;
+        this.param2 = param2;
+    }
+
+    public int getColor2(double percent) {
+
+        // ugly code - given an input percentage (0.0-1.0) this will produce a colour from a "wide rainbow"
+        // from purple (low) to red(high).  This is useful for producing value-based colourations (e.g., altitude)
+
+        double a = (1. - percent) * 5.;
+        int X = (int)Math.floor(a);
+        int Y = (int)(Math.floor(255 * (a - X)));
+        int r = 0,g = 0,b = 0;
+        switch (X) {
+            case 0:
+                r = 255;
+                g = Y;
+                b = 0;
+                break;
+            case 1:
+                r = 255 - Y;
+                g = 255;
+                b = 0;
+                break;
+            case 2:
+                r = 0;
+                g = 255;
+                b = Y;
+                break;
+            case 3:
+                r = 0;
+                g = 255 - Y;
+                b = 255;
+                break;
+            case 4:
+                r = Y;
+                g = 0;
+                b = 255;
+                break;
+            case 5:
+                r = 255;
+                g = 0;
+                b = 255;
+                break;
+        }
+        return 0xFF000000 + (r<<16) + (g<<8) + b;
+    }
+
+
+    private List<GPXUtilities.WptPt> resampleAltitude(List<GPXUtilities.WptPt> pts, double dist) {
+
+        List<GPXUtilities.WptPt> track = resampleTrack(pts, dist);
+
+        int halfC = getColor2(0.5);
+
+        // Calculate the absolutes of the altitude variations
+        Double max = Double.NEGATIVE_INFINITY;
+        Double min = Double.POSITIVE_INFINITY;
+        for (GPXUtilities.WptPt pt : track) {
+            max = Math.max(max, pt.ele);
+            min = Math.min(min, pt.ele);
+            pt.colourARGB = halfC;
+        }
+        Double elevationRange = max-min;
+        if (elevationRange > 0)
+            for (GPXUtilities.WptPt pt : track)
+                pt.colourARGB = getColor2((pt.ele - min)/elevationRange);
+
+        return track;
+    }
+
+
+    private List<GPXUtilities.WptPt> resampleSpeed(List<GPXUtilities.WptPt> pts, double dist) {
+
+        List<GPXUtilities.WptPt> track = resampleTrack(pts, dist);
+
+        GPXUtilities.WptPt lastPt = track.get(0);
+        lastPt.speed = 0;
+
+        // calculate speeds
+        for (int i=1; i<track.size(); i++) {
+            GPXUtilities.WptPt pt = track.get(i);
+            double delta = pt.time - lastPt.time;
+            if (delta>0)
+                pt.speed = MapUtils.getDistance(pt.getLatitude(),pt.getLongitude(),
```

Comment:
As pointed earlier please try to put curve braces around if/else/for. Thanks in advance.